### PR TITLE
TGS3/TGS4 Support! Tested and confirmed working!

### PR DESCRIPTION
**Pull request by Couls**
**What does this PR do:**
The server is now compatible with TGS3 out of the box, it's working in my tests, TGS3 allows for the ability to testmerge multiple PRs and also allows support for a discord bot that notifies a channel about testmerges in progress. Automatic changelog updating is also included with the suite! Meaning no more manual changes.

Refer to this guide for setting it up:https://github.com/tgstation/tgstation-server/blob/V3/README.md

Screenshots of it working:
image
image
image

There is a new verb in the OOC tab that shows you the BYOND version of the server, the current commit and the current testmerges. This also updates rust_g to the latest version.

**Changelog:**
:cl:
add: Paradise now supports TGS3 and TGS4
/:cl:

TGS4 is also supported by this PR and is much easier to set up and configure! Guide here: https://github.com/tgstation/tgstation-server/blob/master/README.md

### CODE REVIEW

**Comment by AffectedArc07** [comment]
Might wanna add the DMAPI for TGS4, since the API is a requirement as opposed to optional in TGS3. 

Currently we /can/ run on TGS3 since I added that a while back, just just finished my half done job. Think I also had a DMAPI PR up at some point as well.

File: TGS3.json [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
+    "synchronize_paths": [
+      "html/changelog.html",
+      "html/changelogs/*"
+    ],
     "static_directories": [
       "config",
       "data"
+    ],
+    "dlls": [
+      "libmysql.dll"
```
Comment (AffectedArc07) [inline TGS3.json]: Add `rust_g.dll` here also

**Comment by AffectedArc07** [comment]
Know what ill test API verification on TGS4 since I have that running in production

**Comment by AffectedArc07** [comment]
image
image
API appears to work on TGS4, since TGS4 fails deployment if it cannot verify the API when starting the server.

**Comment by taukausanake** [comment]
That bot needs more _HECK_ in it. This is neat though. I don't know how to use it but I think I'd like it if I did

File: TGS3.json (hunk 1 above)
Comment (Couls) [reply TGS3.json]: Works fine without it and TG removed it from their json a while back, probably for the best if someone wants to update the dll or stuff, didn’t cause any problems in my tests

**Comment by Couls** [comment]
It working with TGS4 is really good news as the interface is much better

**Comment by Allfd** [comment]
How much testing have you done on this?

Some of the TGS feature set is not implemented on paradise, things like map control and a few others.  How gracefully does TGS4/Paracode fail when they use a non-implimented API call?

**Comment by Couls** [comment]
There are no non-implemented API calls in our code, we have the basic version of the API that's been converted over to normal functions, things like map control and the such aren't actually handled by TGS4 as far as I know, DMAPI would not work if any of the API calls weren't defined so I assume there won't be any issues at all, testmerging deploying and merging and the such was all tested by me on my branch and repository, the changelog also compiles now

File: code/__DEFINES/rust_g.dm [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// rust_g.dm - DM API for rust_g extension library
+#define RUST_G "rust_g"
```
Comment (AffectedArc07) [inline code/__DEFINES/rust_g.dm]: We already have this defined
https://github.com/ParadiseSS13/Paradise/blob/master/code/__HELPERS/_logging.dm#L2

File: code/__DEFINES/rust_g.dm [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+#define rustg_dmi_strip_metadata(fname) call(RUST_G, "dmi_strip_metadata")(fname)
+
+#define rustg_git_revparse(rev) call(RUST_G, "rg_git_revparse")(rev)
+#define rustg_git_commit_date(rev) call(RUST_G, "rg_git_commit_date")(rev)
+
+#define rustg_log_write(fname, text) call(RUST_G, "log_write")(fname, text)
+/proc/rustg_log_close_all() return call(RUST_G, "log_close_all")()
```
Comment (AffectedArc07) [inline code/__DEFINES/rust_g.dm]: We already have this
https://github.com/ParadiseSS13/Paradise/blob/master/code/__HELPERS/_logging.dm#L162

File: code/__DEFINES/rust_g.dm [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// rust_g.dm - DM API for rust_g extension library
+#define RUST_G "rust_g"
+
+#define rustg_dmi_strip_metadata(fname) call(RUST_G, "dmi_strip_metadata")(fname)
+
+#define rustg_git_revparse(rev) call(RUST_G, "rg_git_revparse")(rev)
+#define rustg_git_commit_date(rev) call(RUST_G, "rg_git_commit_date")(rev)
+
+#define rustg_log_write(fname, text) call(RUST_G, "log_write")(fname, text)
```
Comment (AffectedArc07) [inline code/__DEFINES/rust_g.dm]: We already have this
https://github.com/ParadiseSS13/Paradise/blob/master/code/__HELPERS/_logging.dm#L15

File: code/datums/datumvars.dm [hunk 5]
```diff
@@ -1,5 +1,8 @@
 // reference: /client/proc/modify_variables(var/atom/O, var/param_var_name = null, var/autodetect_class = 0)
 
+/datum/proc/CanProcCall(procname)
```
Comment (AffectedArc07) [inline code/datums/datumvars.dm]: Shouldnt this be added in a different PR

File: code/datums/helper_datums/getrev.dm [hunk 6]
```diff
@@ -0,0 +1,77 @@
+/datum/getrev
```
Comment (AffectedArc07) [inline code/datums/helper_datums/getrev.dm]: This may runtime on this code depending on if Kyet has us running in an isolated container or a full git repo. Check with him before doing this

File: code/game/world.dm [hunk 7]
```diff
@@ -417,6 +421,8 @@ var/failed_old_db_connections = 0
 		fcopy(GLOB.config_error_log, "[GLOB.log_directory]/config_error.log")
 		fdel(GLOB.config_error_log)
 
+	log_runtime(GLOB.revdata.get_log_message())
```
Comment (AffectedArc07) [inline code/game/world.dm]: This should be a `log_world()` in an ideal world
